File: q0872/q0872.java
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        // Call DFS to get the leaves of tree 1:
        List<Integer> tree1Leaves = new ArrayList<>();
        dfsHelper(root1, tree1Leaves);

        // Call DFS to get the leaves of tree 2:
        List<Integer> tree2Leaves = new ArrayList<>();
        dfsHelper(root2, tree2Leaves);

        // Compare if the leaves are the same
        return tree1Leaves.equals(tree2Leaves);
    }

    // dfsHelper to get all the leave values
    private void dfsHelper(TreeNode node, List<Integer> leaves) {
        if (node == null)
            return;
        if (node.left == null && node.right == null) {
            leaves.add(node.val);
            return;
        }
        dfsHelper(node.left, leaves);
        dfsHelper(node.right, leaves);
    }
}

File: q0872/q0872.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        tree1Leaves, tree2Leaves = [], []
        self.dfsHelper(root1, tree1Leaves)
        self.dfsHelper(root2, tree2Leaves)
        return tree1Leaves == tree2Leaves
    

    def dfsHelper(self, root, leaves):
        if not root: return
        if not root.left and not root.right: 
            leaves.append(root.val)
            return 
        self.dfsHelper(root.left, leaves)
        self.dfsHelper(root.right, leaves)